xdsl/frontend/passes/promote.py

```diff
@@ -0,0 +1,320 @@
+
+            # Start by finding updates to this symbol in both true and false
+            # blocks.
+            def promote_from_block(block: Block) -> symref.Update:
+                for op in block.ops:
+                    if isinstance(op, symref.Update):
+                        if symbol == op.attributes["symbol"].data.data:
+                            return op
+                return None
+
+            true_block_update_op = promote_from_block(true_block)
+            false_block_update_op = promote_from_block(false_block)
+
+            # No updates to this symbol - move on.
+            if true_block_update_op is None and false_block_update_op is None:
+                continue
+
+            # Record the type of the update to reconstruct a new scf.if later.
+            if true_block_update_op is not None:
+                update_ty = true_block_update_op.operands[0].typ 
+            else:
+                update_ty = false_block_update_op.operands[0].typ
+
+            # Otherwise there is an update. First, we check if some of the
+            # blocks haven't updated the symbol and it was also not fetched. In
+            # this case, we need to introduce a fetch operation in the parent
+            # block.
+            if (true_block_update_op is None or false_block_update_op is None
+                ) and promoted_fetch_ops[i] is None:
+                new_fetch_op = symref.Fetch.get(symbol, update_ty)
+                insert_before(if_op, new_fetch_op)
+                promoted_fetch_ops[i] = new_fetch_op
+
+            def promote_update(update_op: None | symref.Update,
+                               yield_operands: List[SSAValue]) -> SSAValue:
+                if update_op is not None:
+                    yield_value = update_op.operands[0]
+                    rewriter.erase_op(update_op)
+                else:
+                    yield_value = promoted_fetch_ops[i].results[0]
+                yield_operands.append(yield_value)
+
+        # Promote updates by making them operands to scf.yield.
+        promoted_symbols.append(symbol)
+        promote_update(true_block_update_op, true_block_yield_operands)
+        promote_update(false_block_update_op, false_block_yield_operands)
+
+        # Next, actually construct a new scf.if operation.
+        true_block_yield_operands = list(
+            true_block.ops[-1].operands) + true_block_yield_operands
+        false_block_yield_operands = list(
+            false_block.ops[-1].operands) + false_block_yield_operands
+
+        rewriter.replace_op(true_block.ops[-1],
+                            scf.Yield.get(*true_block_yield_operands))
+        rewriter.replace_op(false_block.ops[-1],
+                            scf.Yield.get(*false_block_yield_operands))
+        return_types = list(map(lambda op: op.typ, true_block_yield_operands))
+
+        new_true_region = Region()
+        if_op.true_region.clone_into(new_true_region)
```

Comment:
There is also [this method](https://github.com/xdslproject/xdsl/blob/main/xdsl/rewriter.py#L145) in the Rewriter. should do the same thing.